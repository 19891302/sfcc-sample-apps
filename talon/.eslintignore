**/node_modules/*
**/dist/*
target
coverage
**/__tests__/fixtures/*