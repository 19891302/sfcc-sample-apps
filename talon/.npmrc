# We have two registries we could use:
# Internal (private)
registry=https://nexus.soma.salesforce.com/nexus/content/groups/npm-all/