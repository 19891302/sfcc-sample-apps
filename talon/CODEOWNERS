# Global Code Owner
#   @zachary-hawtof @nrobertdehault @tconn @nwong