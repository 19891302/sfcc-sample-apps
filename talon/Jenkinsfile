/* Before using this example, check if the below release tag(v0.3.0) is the latest.
 * You can find the latest release tag at https://git.soma.salesforce.com/dci/sfci-pipeline-sharedlib/releases
 */
@Library('sfci-pipeline-sharedlib@master') _

/* import is required only if you use one of the classes specified here:
 * https://git.soma.salesforce.com/dci/sfci-pipeline-sharedlib#classes
 */
import net.sfdc.dci.BuildUtils
import net.sfdc.dci.GitHubUtils
import groovy.json.JsonOutput

/* Specify the build image you want to use.
 * This build image uses https://git.soma.salesforce.com/communities/centos-sfci-nodejs-puppeteer
 * Not specifying a build image will use the default build image specified here:
 * https://git.soma.salesforce.com/dci/sfci-pipeline-sharedlib/blob/master/vars/executePipeline.groovy#L4
 */
def envDef = [ buildImage: 'ops0-artifactrepo1-0-prd.data.sfdc.net/communities/centos-sfci-nodejs-puppeteer:latest' ]
def stagingProfileId = '58da552823cd6'

// define any release branches here
env.RELEASE_BRANCHES = ['220', '218', '216']

// The checked-out revision / commit information.
// We use this elsewhere and leak it here for convenience.
gitRepo = null
commitSha = null

executePipeline(envDef) {

    // Checkout the commit and extract some useful values.
    echo 'Checking out git repo...'
    stage('Checkout') {
        revision = checkout scm
        gitRepo = revision.GIT_URL
        commitSha = revision.GIT_COMMIT
    }

    stage('NPM') {
        stage('NPM Init') {
            npmInit(null)
        }
        stage('NPM Install') {
            sh 'npm_config_tarball=/usr/lib/node-v10.13.0-headers.tar.gz yarn install'
        }
        stage('NPM Build') {
            sh 'yarn build'
        }        
        stage('Code Validation') {
            parallel([
                failFast: true,
                "Code Conformance": {
                    reportStepsToCurrentCommitGitHubStatus('Code Conformance') {
                        sh 'yarn lint'
                    }
                },
                "Jest Tests": {
                    reportStepsToCurrentCommitGitHubStatus('Code Coverage') {
                        sh 'yarn coverage'
                    }
                },
                "Integration Tests": {
                    reportStepsToCurrentCommitGitHubStatus('Integration Tests') {
                        sh 'yarn test:integration'
                    }
                }
            ])
        }
        stage('Collect Reports') {
            publishHTML([
                allowMissing: false,
                alwaysLinkToLastBuild: false,
                keepAll: false,
                reportDir: 'jest-report',
                reportFiles: 'index.html',
                reportName: 'JEST Results',
                reportTitles: ''
            ])
            publishHTML([
                allowMissing: false,
                alwaysLinkToLastBuild: false,
                keepAll: false,
                reportDir: 'coverage/lcov-report',
                reportFiles: 'index.html',
                reportName: 'JEST Coverage',
                reportTitles: ''
            ])
        }  
    }

    stage('Maven') {
        stage('Initialize') {
            mavenInit()
        }  
        if (BuildUtils.isPullRequestBuild(env) || BuildUtils.isReleaseBuild(env)) {
            stage('Build') {
                reportStepsToCurrentCommitGitHubStatus('Maven Build') {
                    withEnv(["COMPILE_LWC_COMPONENTS=true"]) {
                        mavenVersionsSet([managed : true])
                        mavenBuild()
                    }
                }
            }
            // If this is a release build, we can stage and release the artifacts.
            if (BuildUtils.isReleaseBuild(env)) {
                stage('Stage Artifacts') {
                    reportStepsToCurrentCommitGitHubStatus('Artifact Staging') {
                        mavenStageArtifacts([staging_profile_id : stagingProfileId])
                    }
                }
            //  TODO: reenable when itests back running, and fewer breaking changes :( 
            //    stage('Trigger Core Itests') {
            //        env.AUTOBUILD_ENV = 'prod'
            //        reportStepsToCurrentCommitGitHubStatus('Core Integration Tests') {
            //            triggerCoreItests()
            //        }
            //    }
                stage('Release') {
                    reportStepsToCurrentCommitGitHubStatus('Artifact Promotion to Release') {
                        mavenPromoteArtifacts()
                    }
                }
                stage('GUS Compliance'){
                    git2gus()
                }
            }
        } else {
            mavenBuild()
        }
    }
}

/**
 *  Reports the progress and outcome of one or more steps to GitHub as the status of a GitHub Status API context.
 *  The specified GitHub Status API context will be marked as "pending" before the steps are executed.
 *  If the steps throw an error, the context will be marked as "failure"; otherwise, it will be marked as "success".
 *
 *  @param  gitHubStatusApiContextName
 *      The name of the GitHub Status API context to which the progress of the steps are reported.
 *
 *  @param  steps
 *      The build steps to execute.
**/
void reportStepsToCurrentCommitGitHubStatus(String gitHubStatusApiContextName, Closure steps) {

    setCurrentCommitStatus([state: 'pending', context: gitHubStatusApiContextName])
    try {
        steps()
        setCurrentCommitStatus([state: 'success', context: gitHubStatusApiContextName])
    }
    catch (e) {
        setCurrentCommitStatus([state: 'failure', context: gitHubStatusApiContextName, description: e.getMessage()])
        
        // We are just spying on any errors for reporting purposes, so re-throw.
        throw e
    }
}

/**
 *  Sets the GitHub API Status for the current (ambient) commit.
 *
 *  @param  status
 *      A set of GitHub Status API parameters that specify the status of the commit, as supported by https://developer.github.com/v3/repos/statuses/#create-a-status.
 **/
void setCurrentCommitStatus(status) {

    // If no explicit target_url has been set, use this build.
    status.target_url = status.target_url ?: env.BUILD_URL
    setCommitStatus(gitRepo, commitSha, status)
}

/**
 *  Sets the status of a GitHub commit.
 *
 *  @param  repositoryUri
 *      The URI of the GitHub repository (with or without a ".git" suffix) to which the status is reported.
 *  @param  commitSha
 *      The commit (e.g. SHA1) to which the status is associated.
 *  @param  status
 *      The status to set for the commit, as supported by https://developer.github.com/v3/repos/statuses/#create-a-status.
 *
 *  @returns
 *      The HTTP response of the status operation.
**/
void setCommitStatus(repositoryUri, commitSha, status) {

    // Extract the host, org, and repo from the commit checkout.
    // We'll find this in repositoryUri, which looks like "https://git.soma.salesforce.com/FeedsUI/ui-feeds-components.git"
    String[] repositoryUriParts = repositoryUri.split("/")

    String host = repositoryUriParts[2]
    String org = repositoryUriParts[3]
    String repo = repositoryUriParts[4].replace(".git", "") // Remove any trailing ".git" repository specifier
    String credentialsId = "sfci-git"   // Set by Jenkins (see https://git.soma.salesforce.com/dci/sfci-pipeline-sharedlib for an example)
    
    // Use the utility to form up our base URI to GitHub.
    String repositoryBaseUri = GitHubUtils.apiV3Repos(host, org, repo)

    // Use the GitHub Status API endpoint for this specific commit.
    String commitStatusUri = "${repositoryBaseUri}/statuses/${commitSha}"

    // Override the default content and accept types of application/json
    // by specifying NOT_SET and providing them as custom headers
    customHeaders = GitHubUtils.previewCustomHeaders(this, credentialsId)

    // POST the status to GitHub
    doPost(commitStatusUri, status, customHeaders, null, null)
}

/**
 * This method is stolen from BuildUtils.doPost, but modified to properly support parallel build steps.
 *
 * This method performs a HTTP POST request on requested resource and returns the response as a groovy map.
 * customHeaders is an array of maps, accepted keys as 'name' - Name of the header, 'value' - Value of the header
 * postBody can be a map or an array or any groovy primitive type, the method will convert it into a valid json string for the request.
 * Ex: [[name: 'Authorization', value: 'token sometoken'], [name: 'Accept', value: 'application/vnd.github.v3+json']]
 * The method produces a response of type groovy map, with key 'status' - response status code from the api request, 'content' - json response as groovy map
 *
 * Note: to override the {@code Content-Type} or {@code Accept} headers you
 * must set the corresponding {@code contentType} or {@code acceptType}
 * parameters to 'NOT_SET' then override them in {@code customHeaders}.
 *
 * For the supported types, refer to:
 * http://javadoc.jenkins-ci.org/http_request/jenkins/plugins/http_request/MimeType.html
**/
def doPost(def url, def postBody, def customHeaders, def contentType, def acceptType) {
    def bodyAsString = JsonOutput.toJson(postBody)
    def apiResponse = httpRequest(
            url: url,
            httpMode: 'POST',
            customHeaders: customHeaders ?: [],
            contentType: contentType ?: 'APPLICATION_JSON',
            acceptType: acceptType ?: 'APPLICATION_JSON',
            requestBody: bodyAsString
            )
    def json = readJSON([text: apiResponse.content])
    def response = [status: apiResponse.status, content: json]
    return response
}
