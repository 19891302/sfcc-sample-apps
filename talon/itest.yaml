core:
  # codeline of //app/main/core to test against
  codeline: main
  # core modules that need to be updated on the core pom/xml
  modules:
    ui-communities-talon:
      optional_history_file: null
      # maven property name for the module's version
      properties_name: ui-communities-talon.version
  # execute tests that are part of the ftest labels
  ftest_labels: [~ignore,talon.itest]
  # execute both A and B run,
  # where A run is with your staged artifact
  #   and B run is without updating core pom.xml with the version of your artifact
  ab_enabled: true

  # this is a timeout for SFCI to stop polling
  # unit is minutes
  # pending is the time SFCI waits when autobuild is waiting to be allocated resource
  # running is the time SFCI waits when itests are in-progress or RUNNING state
  timeout:
    pending: 90
    running: 240