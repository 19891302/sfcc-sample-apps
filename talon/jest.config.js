module.exports = {
    projects: [
        "packages/talon-common",
        "packages/talon-compiler",
        "packages/talon-framework",
        "packages/talon-template-flashhelp"
    ],
    coverageThreshold: {
        "global": {
            "statements": 90,
            "branches": 75,
            "functions": 90,
            "lines": 90
        }
    },
    collectCoverageFrom: [
        "**/src/**/*.{js}",
        "!**/node_modules/**",
        "!**/src/index.js",
        '!**/src/modules/talon/intlLibrary/**',
        '!**/src/modules/talon/auraInstrumentation/**',
        '!**/src/modules/community_flashhelp/form/**',
        "!**/__tests__/**",
        "!**/jest-modules/**"
    ],
    testResultsProcessor: "<rootDir>/node_modules/jest-html-reporter",
    testURL: "https://localhost",
    testRegex: "/__tests__/.*.test.js$"
};