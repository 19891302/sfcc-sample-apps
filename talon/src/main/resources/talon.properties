pom.version=${pom.version}
project.artifactId=${project.artifactId}