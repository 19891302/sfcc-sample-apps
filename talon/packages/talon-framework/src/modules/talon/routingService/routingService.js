import { getBasePath } from "talon/configProvider";
import { assert, autoBind, getQueryStringParameterByName } from "talon/utils";
import { getThemeLayoutByView } from "talon/themeService";

const WHITELISTED_PARAMETERS = ['mode'];

/**
 * Routing service class.
 *
 * A single instance of it will be used throughout the app
 * and selected methods will be exported.
 *
 * We still export the class itself for testing purpose so that we can
 * create as many instances as needed.
 */
export class RoutingService {
    router;
    routesByName = {};
    observers = new Set();
    currentRoute;

    /**
     * Register a set of routes and start the router.
     *
     * See JSON schema at https://salesforce.quip.com/EDbbAKBI92ZM#LeeACA0AB7l.
     *
     * See validate-routes.js#validateRoutes() for the validation rules.
     *
     * @param {Object[]} routes - The routes to register
     */
    registerRoutes(routes) {
        assert(this.router, "Router implementation not set");

        routes.forEach((route) => {
            // save route for lookup by name
            this.routesByName[route.name] = route;

            const callback = this.onRouteChange.bind(this, route.name);

            if (route.isDefault) {
                this.router('*', callback);
            } else {
                this.router(route.path, callback);
            }
        });

        // set base path
        const basePath = getBasePath();
        if (basePath) {
            this.router.base(basePath);
        }

        // start the router
        this.router.start();
    }

    /**
     * Get the URL for a given route.
     * This can be used to populate links href.
     *
     * The returned URL will include any whitelisted parameters
     * specified in the current URL.
     *
     * TODO handle route parameters
     *
     * @param {string} name - The route name
     */
    getRouteUrl(name) {
        assert(this.routesByName[name], `Unknown route: ${name}`);

        const routeUrl = getBasePath() + (this.routesByName[name] || {}).path;

        // add present whitelisted parameters to the route URL
        const params = WHITELISTED_PARAMETERS.reduce((acc, param) => {
            const value = getQueryStringParameterByName(param);
            return (value && acc.concat([param + '=' + encodeURIComponent(value)])) || acc;
        }, []);

        if (Object.keys(params).length > 0) {
            return routeUrl + '?' + params.join('&');
        }

        return routeUrl;
    }

    /**
     * Returns a route which has been registered
     *
     * @param {string} name
     */
    getRoute(name) {
        assert(this.routesByName[name], `Unknown route: ${name}`);
        return this.routesByName[name] || {};
    }

    /**
     * Callback invoked by the router when route has changed.
     *
     * This will either reload the whole page if the current and target
     * theme layouts are different, or notify the observers if not.
     *
     * @param {string} name the target route name
     */
    onRouteChange(name) {
        const route = this.routesByName[name];

        // the app container and the router container so that it can create and render
        // the page component and the theme layout
        this.currentRoute = route;
        this.themeLayout = getThemeLayoutByView(route.view);
        this.observers.forEach(observer => {
            observer.next(route);
        });
    }

    /**
     * Navigates to the route with the given name.
     *
     * TODO handle route parameters
     *
     * @param {string} name - The route name
     */
    navigateToRoute(name, params) {
        let path;
        const route = this.getRouteUrl(name).split('/:');

        path = route[0];

        if ( params && route.length > 0 ) {
            Object.keys(params).forEach((key) => {
                path += `/${params[key]}`;
            });
        }

        this.router.show(path);
    }

    /**
     * Subscribe to route changes.
     *
     * @param {Object} observer - An observer that will be notified on route change.
     *                      Observers are objects with a next() method that will be invoked
     *                      with the new route every time the route changes.
     * @returns A subscription object with an unsubscribe() method that can be used to stop
     *          receiving notifications.
     * @example
     *  const subscription = subscribe({
     *      next: (route) => {
     *          // handle route change
     *      }
     *  });
     *
     *  // later
     *  subscription.unsubscribe();
     *
     */
    subscribe(observer) {
        if (this.currentRoute) {
            observer.next(this.currentRoute);
        }

        this.observers.add(observer);

        return {
            unsubscribe: () => {
                this.observers.delete(observer);
            }
        };
    }

    /**
     * Sets the 3rd party router, currently https://visionmedia.github.io/page.js/.
     * We do this so that we can compile in core where the 3rd party router is not available.
     *
     * Also resets current routes and observer since they are tied to the current router
     * although setting router impl multiple times should only happen in tests.
     *
     * @param {Object} routerImpl - 3rd party router
     */
    setRouter(routerImpl) {
        this.router = routerImpl;
    }
}

// create an instance with bound methods so that they can be exported
const instance = autoBind(new RoutingService());

export const { registerRoutes, getRouteUrl, getRoute, navigateToRoute, subscribe, setRouter } = instance;

export default { registerRoutes, getRouteUrl, getRoute, navigateToRoute, subscribe, setRouter };