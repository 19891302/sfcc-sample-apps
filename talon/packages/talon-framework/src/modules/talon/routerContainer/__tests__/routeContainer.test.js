import { createElement } from 'lwc';
import RouterContainer from 'talon/routerContainer';
import { subscribe } from 'talon/routingService';
import { default as mockAbout } from './fixtures/aboutTemplate';

jest.mock('talon/moduleRegistry', () => {
    return {
        getTemplate: () => {
            return Promise.resolve(mockAbout);
        }
    };
});

jest.mock('talon/routingService', () => {
    return {
        subscribe: jest.fn()
    };
});

beforeEach(() => {
    while (document.body.firstChild) {
        document.body.removeChild(document.body.firstChild);
    }

    jest.resetAllMocks();

    subscribe.mockReturnValue({
        unsubscribe: jest.fn()
    });
});

describe('talon/routerContainer', () => {
    it('renders no route', async () => {
        const element = createElement('talon-router-container', { is: RouterContainer });
        document.body.appendChild(element);

        expect(element).toMatchSnapshot();
    });

    it('renders route', () => {
        const element = createElement('talon-router-container', { is: RouterContainer });
        document.body.appendChild(element);

        const observer = subscribe.mock.calls[0][0];
        observer.next({ view: "about" });

        return Promise.resolve().then(() => {
            return new Promise(resolve => {
                window.requestAnimationFrame(resolve);
            });
        }).then(() => {
            expect(element).toMatchSnapshot();
        });
    });

    it('subscribes to route changes', async () => {
        const element = createElement('talon-router-container', { is: RouterContainer });
        document.body.appendChild(element);

        expect(subscribe).toHaveBeenCalledTimes(1);
    });

    it('unsubscribes when disconnected', async () => {
        const element = createElement('talon-router-container', { is: RouterContainer });
        document.body.appendChild(element);
        document.body.removeChild(element);

        const { unsubscribe } = subscribe.mock.results[0].value;
        expect(unsubscribe).toHaveBeenCalledTimes(1);
    });
});
