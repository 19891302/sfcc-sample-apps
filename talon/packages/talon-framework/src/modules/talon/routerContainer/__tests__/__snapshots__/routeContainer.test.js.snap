// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`talon/routerContainer renders no route 1`] = `
<talon-router-container>
  #shadow-root(open)
</talon-router-container>
`;

exports[`talon/routerContainer renders route 1`] = `
<talon-router-container>
  #shadow-root(open)
    About
</talon-router-container>
`;
